java/arcs/core/analysis/PredicateUtils.kt

```diff
@@ -15,62 +15,79 @@ import arcs.core.data.InformationFlowLabel
+ *            A and B : {(110, 110)}
+ *       A and not(B) : {(110, 100)}
+ *             A or B : {(100, 100), (010, 010)}
+ *        A or not(B) : {(100, 100), (010, 000)}
+ *   (A and B) or (C) : {(110, 110), (001, 001)}
  */
-fun InformationFlowLabel.Predicate.asSetOfBitSets(
+fun InformationFlowLabel.Predicate.asDNF(
     indices: Map<InformationFlowLabel, Int>
-): Set<BitSet> {
+): Set<Pair<BitSet, BitSet>> {
     // TODO(b/157530728): This is not very efficient. We will want to replace this with
     // an implementation that uses a Binary Decision Diagram (BDD).
     when (this) {
         is Predicate.Label -> {
-            val result = BitSet(indices.size)
-            result.set(requireNotNull(indices[label]))
-            return setOf(result)
+            val index = requireNotNull(indices[label])
+            val result = BitSet(indices.size).apply { set(index) }
+            return setOf(result to result)
         }
         is Predicate.Not -> {
             val labelPredicate = requireNotNull(predicate as? Predicate.Label) {
                 // TODO(b/157530728): It will be easy to handle `not` when we have a proper
                 // datastructure like BDDs. For now, we only support `not` on labels.
                 "Not is only supported for label predicates when converting to bitsets!"
             }
-            val labelIndex = requireNotNull(indices[labelPredicate.label])
-            // Not(A) = empty \/ B \/ C ...
-            return (0..(indices.size - 1))
-                .filter { it != labelIndex }
-                .map { index -> BitSet(indices.size).apply { set(index) } }
-                .plus(BitSet(indices.size))
-                .toSet()
+            val index = requireNotNull(indices[labelPredicate.label])
+            val mask = BitSet(indices.size).apply { set(index) }
+            val result = BitSet(indices.size)
+            return setOf(mask to result)
         }
         is Predicate.Or -> {
-            val lhsBitSets = lhs.asSetOfBitSets(indices)
-            val rhsBitSets = rhs.asSetOfBitSets(indices)
+            val lhsBitSets = lhs.asDNF(indices)
+            val rhsBitSets = rhs.asDNF(indices)
             return lhsBitSets union rhsBitSets
         }
         is Predicate.And -> {
-            val lhsBitSets = lhs.asSetOfBitSets(indices)
-            val rhsBitSets = rhs.asSetOfBitSets(indices)
-            val result = mutableSetOf<BitSet>()
-            lhsBitSets.forEach { lhsBitSet ->
-                rhsBitSets.forEach { rhsBitSet ->
-                    val combined = BitSet(indices.size)
-                    combined.or(lhsBitSet)
-                    combined.or(rhsBitSet)
-                    result.add(combined)
+            val lhsBitSets = lhs.asDNF(indices)
+            val rhsBitSets = rhs.asDNF(indices)
+            val result = mutableSetOf<Pair<BitSet, BitSet>>()
+            // Apply distributivity law.
+            lhsBitSets.forEach { (lhsMask, lhsBitSet) ->
```

Comment:
Probably worth pulling this into its own function.

Reply:
Done.